File: ios/BockMedia/Domain/PlaybackFocus.swift
import Foundation

/// Remembers which speaker the user last started playback on.
enum PlaybackFocus {
    private static let groupPrefix = "group:"

    private(set) static var pendingDeviceValue: String?
    private(set) static var pendingDeviceLabel: String?
    private(set) static var focusedDeviceId: String?
    private(set) static var generation = 0

    static func notePlayStarted(_ deviceValue: String, label: String?) {
        pendingDeviceValue = deviceValue.trimmingCharacters(in: .whitespaces).nilIfBlank
        pendingDeviceLabel = label?.trimmingCharacters(in: .whitespaces).nilIfBlank
        generation += 1
    }

    @discardableResult
    static func syncPendingFocus(items: [NowPlayingDeviceItem], alexaDevices: [AlexaDevice] = []) -> Bool {
        guard let deviceValue = pendingDeviceValue else { return false }
        guard let match = findMatchingItem(items: items, deviceValue: deviceValue, label: pendingDeviceLabel, alexaDevices: alexaDevices) else {
            return false
        }
        focusedDeviceId = match.deviceId
        pendingDeviceValue = nil
        pendingDeviceLabel = nil
        generation += 1
        return true
    }

    static func focusDeviceId(_ deviceId: String) {
        focusedDeviceId = deviceId
        pendingDeviceValue = nil
        pendingDeviceLabel = nil
        generation += 1
    }

    static func resolveFocusedItem(items: [NowPlayingDeviceItem], alexaDevices: [AlexaDevice] = []) -> NowPlayingDeviceItem? {
        guard !items.isEmpty else { return nil }
        _ = syncPendingFocus(items: items, alexaDevices: alexaDevices)
        if let id = focusedDeviceId, let item = items.first(where: { $0.deviceId == id }) {
            return item
        }
        return items.first(where: { !$0.paused }) ?? items.first
    }

    private static func findMatchingItem(
        items: [NowPlayingDeviceItem],
        deviceValue: String,
        label: String?,
        alexaDevices: [AlexaDevice]
    ) -> NowPlayingDeviceItem? {
        if let label, let match = matchByName(items: items, name: label) { return match }

        if !deviceValue.hasPrefix(groupPrefix) {
            if let alexaName = alexaDevices.first(where: { $0.serial == deviceValue })?.name,
               let match = matchByName(items: items, name: alexaName) {
                return match
            }
        }
        return nil
    }

    private static func matchByName(items: [NowPlayingDeviceItem], name: String) -> NowPlayingDeviceItem? {
        if let exact = items.first(where: { ($0.deviceName ?? "").caseInsensitiveCompare(name) == .orderedSame }) {
            return exact
        }
        return items.first { item in
            guard let deviceName = item.deviceName else { return false }
            return deviceName.localizedCaseInsensitiveContains(name) || name.localizedCaseInsensitiveContains(deviceName)
        }
    }
}

private extension String {
    var nilIfBlank: String? {
        let t = trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
    }
}

File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/PlaybackFocus.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.AlexaDevice
import com.bockmedia.console.data.api.dto.NowPlayingDeviceItem
import com.bockmedia.console.media.LOCAL_PHONE_DEVICE_ID
import com.bockmedia.console.media.isLocalPhoneDevice

/** Remembers which speaker the user last started playback on. */
object PlaybackFocus {
    @Volatile
    var pendingDeviceValue: String? = null
        private set

    @Volatile
    var pendingDeviceLabel: String? = null
        private set

    @Volatile
    var focusedDeviceId: String? = null
        private set

    @Volatile
    var generation: Int = 0
        private set

    fun notePlayStarted(deviceValue: String, deviceLabel: String?) {
        if (deviceValue == LOCAL_PHONE_DEVICE_ID || isLocalPhoneDevice(deviceValue)) {
            focusedDeviceId = LOCAL_PHONE_DEVICE_ID
            pendingDeviceValue = null
            pendingDeviceLabel = null
            generation++
            return
        }
        pendingDeviceValue = deviceValue.trim().takeIf { it.isNotEmpty() }
        pendingDeviceLabel = deviceLabel?.trim()?.takeIf { it.isNotEmpty() }
        generation++
    }

    fun clear() {
        pendingDeviceValue = null
        pendingDeviceLabel = null
        focusedDeviceId = null
    }

    fun syncPendingFocus(items: List<NowPlayingDeviceItem>, alexaDevices: List<AlexaDevice> = emptyList()): Boolean {
        val deviceValue = pendingDeviceValue ?: return false
        val label = pendingDeviceLabel

        val match = findMatchingItem(items, deviceValue, label, alexaDevices) ?: return false
        focusedDeviceId = match.deviceId
        pendingDeviceValue = null
        pendingDeviceLabel = null
        generation++
        return true
    }

    fun resolveFocusedItem(
        items: List<NowPlayingDeviceItem>,
        alexaDevices: List<AlexaDevice> = emptyList(),
    ): NowPlayingDeviceItem? {
        if (items.isEmpty()) return null
        syncPendingFocus(items, alexaDevices)
        focusedDeviceId?.let { id -> items.find { it.deviceId == id }?.let { return it } }
        return items.firstOrNull { !it.paused } ?: items.firstOrNull()
    }

    fun focusedIndex(items: List<NowPlayingDeviceItem>): Int {
        if (focusedDeviceId == LOCAL_PHONE_DEVICE_ID) {
            val idx = items.indexOfFirst { isLocalPhoneDevice(it.deviceId) }
            if (idx >= 0) return idx
        }
        val id = focusedDeviceId ?: return 0
        return items.indexOfFirst { it.deviceId == id }.coerceAtLeast(0)
    }

    private fun findMatchingItem(
        items: List<NowPlayingDeviceItem>,
        deviceValue: String,
        label: String?,
        alexaDevices: List<AlexaDevice>,
    ): NowPlayingDeviceItem? {
        label?.let { matchByName(items, it) }?.let { return it }

        if (!deviceValue.startsWith(GROUP_PREFIX)) {
            alexaDevices.firstOrNull { it.serial == deviceValue }?.name?.let { alexaName ->
                matchByName(items, alexaName)?.let { return it }
            }
        }

        return null
    }

    private fun matchByName(items: List<NowPlayingDeviceItem>, name: String): NowPlayingDeviceItem? {
        items.find { it.deviceName.equals(name, ignoreCase = true) }?.let { return it }
        return items.find { item ->
            val deviceName = item.deviceName ?: return@find false
            deviceName.contains(name, ignoreCase = true) || name.contains(deviceName, ignoreCase = true)
        }
    }

    private const val GROUP_PREFIX = "group:"
}
